[Zheng Yilin] iP

yilinzyl opened the pull request:
## Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:

- [X] Managing tasks
- [X] Managing deadlines (coming soon)
- [X] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

yilinzyl committed:
Level-1

yilinzyl committed:
Level-2

yilinzyl committed:
Level-3

yilinzyl committed:
Level-4

yilinzyl committed:
A-TextUiTesting

yilinzyl committed:
Level-5

yilinzyl committed:
Level-6

yilinzyl committed:
Level-7

yilinzyl committed:
Level-8

yilinzyl committed:
Merge branch 'branch-Level-8'

yilinzyl committed:
A-MoreOOP

yilinzyl committed:
A-Packages

yilinzyl committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

yilinzyl committed:
A-Gradle

yilinzyl committed:
A-JUnit

yilinzyl committed:
A-JavaDoc

yilinzyl committed:
Fix style

yilinzyl committed:
Fix coding standard errors

yilinzyl committed:
Resolve merge conflict

yilinzyl committed:
Add some stucture for finding results

yilinzyl committed:
Add find functionality

yilinzyl committed:
Merge find function

vantemoon commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,57 @@
+     */
+    public void run() {
+        ui.showWelcome();
+        boolean isExit = false;
```
I like how you named the boolean variable in this way to make it sounds like a boolean 👍

vantemoon commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,140 @@
+    public static Command parse(String input) throws DukeException {
+        CommandWords commandWord = getCommand(input);
+        switch (commandWord) {
+        case BYE:
```
I like how you noticed to not indent the case clauses. Nice!

vantemoon commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    /**
+     * Display list of found tasks.
```
Should this be "Displays"?

vantemoon commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    /**
+     * Get the next line of user input.
```
Should this be "Gets" or "Returns"?

vantemoon reviewed (commented):
Overall clean and neat! Just a few minor mistakes with the Javadoc comments:)

joosxi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,140 @@
+
+    private static int getTaskNumber(int commandLength, String input) throws DukeException {
+        if (input.length() == commandLength || (input.length() == commandLength + 1 &&
+                input.substring(commandLength, commandLength + 1).equals(" "))) {
```
Perhaps these 2 lines could have been broken into multiple simpler boolean values for ease of readability! :)

joosxi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,140 @@
+     */
+    public static Command parse(String input) throws DukeException {
+        CommandWords commandWord = getCommand(input);
+        switch (commandWord) {
```
I like how you extracted the commandWord from the input and decided on the appropriate response based on enumerations and cases!

joosxi commented on src/main/java/duke/CommandWords.java:
```diff
@@ -0,0 +1,8 @@
+ * Commands that the user can use in the Duke Application.
+ */
+public enum CommandWords {
+    BYE, LIST, UNMARK, MARK, TODO, EVENT, DEADLINE, DELETE, FIND
```
I like how you used enumerations here! I actually didn't think of it myself but I think it's a clever use here! :)

joosxi reviewed (commented):
Hi Yilin! I like how your code was written and I think the effort you spent in creating so many classes in command to parse your code is really commendable! Your code gave me new insights as to how the IP could have been structured in terms of the classes and methods!

yilinzyl committed:
Level-10

yilinzyl committed:
Add GUI to Duke

yilinzyl committed:
Add assertions

Code has no assertions.

Bugs in the code may not be detected and terminated early, which is not
safe.

Let's add some assertions to check assumptions about the program state.

Assertions are used as these checks are performed within functional
code. The runtime can verify them early which is important as assertion
failures indicates a possible bug.

yilinzyl committed:
Merge pull request #2 from yilinzyl/branch-A-Assertions

Add assertions

yilinzyl committed:
Improve code quality

There are some unclear conditionals and no default branch for if-else
statement. There are also long methods with repeated code doing similar
things.

This makes the code difficult to read. It is also prone to error.

Let's
* Redefine conditionals with a name
* Add a default branch and throw an exception if the default branch is
  reached, as it is not supposed to
* Extract repeated code out into methods

Giving boolean conditionals a name makes its purpose more easily
understood. Choosing to still add a default branch even if it is
supposed to be unreachable helps to ensure that the code is written
correctly, and errors should there be an accidentally misuse be easily
detected. Extracting repeated code into methods makes it easy to reuse
the same piece of code for similar purposes.

yilinzyl committed:
Merge pull request #3 from yilinzyl/branch-A-CodeQuality

Improve code quality

yilinzyl committed:
Fix coding standards issues

yilinzyl committed:
Add sort functionality

Order in which tasks are displayed is determined by time in which they
are added.

This is not very friendly to users who want to know what they should
prioritise first. Adding a sort functionality will help users know what
is coming up easily.

Let's
* add a sort function for deadlines alone, events alone, and both
  deadlines and events.

Users might want to view deadlines by when they are due alone. so that
they can start working on these tasks. They might also want to view
sorted events alone to know what is coming up and that they have to
attend. They might also want to view both deadlines and events together
to plan their schedule accordingly.

yilinzyl committed:
Merge pull request #4 from yilinzyl/branch-C-Sort

Add sort functionality

yilinzyl committed:
Add user guide

yilinzyl committed:
Improve UI

yilinzyl committed:
Upload UI screenshot

yilinzyl committed:
Fix user guide hyperlinks

yilinzyl committed:
Rewrite test cases to fit new UI format
